File: backend/app/api/memory.py
from __future__ import annotations

from fastapi import (
    APIRouter,
    HTTPException,
    Query,
)

from backend.app.memory import (
    MemoryWriteOrchestrator,
)


router = APIRouter(
    prefix="/api/memory",
    tags=["memory"],
)


def _orchestrator() -> MemoryWriteOrchestrator:
    return MemoryWriteOrchestrator(
        "data/cybertron.db"
    )


@router.get("/proposals")
async def list_memory_proposals(
    status: str | None = Query(
        default=None,
    ),
    limit: int = Query(
        default=100,
        ge=1,
        le=500,
    ),
) -> dict:
    try:
        proposals = (
            _orchestrator().list_proposals(
                status=status,
                limit=limit,
            )
        )
    except ValueError as exc:
        raise HTTPException(
            status_code=400,
            detail=str(exc),
        ) from exc

    return {
        "status_filter": status,
        "count": len(proposals),
        "proposals": proposals,
    }


@router.get(
    "/proposals/{proposal_id}"
)
async def get_memory_proposal(
    proposal_id: str,
) -> dict:
    orchestrator = _orchestrator()

    proposal = orchestrator.get(
        proposal_id
    )

    if proposal is None:
        raise HTTPException(
            status_code=404,
            detail="Memory proposal not found.",
        )

    return proposal


@router.get(
    "/proposals/{proposal_id}/audit"
)
async def get_memory_proposal_audit(
    proposal_id: str,
) -> dict:
    orchestrator = _orchestrator()

    proposal = orchestrator.get(
        proposal_id
    )

    if proposal is None:
        raise HTTPException(
            status_code=404,
            detail="Memory proposal not found.",
        )

    events = orchestrator.audit_history(
        proposal_id
    )

    return {
        "proposal_id": proposal_id,
        "count": len(events),
        "events": events,
    }


@router.post(
    "/proposals/{proposal_id}/approve"
)
async def approve_memory_proposal(
    proposal_id: str,
) -> dict:
    orchestrator = _orchestrator()

    try:
        proposal = orchestrator.decide(
            proposal_id,
            approved=True,
            actor_type="user",
            actor_id="local-operator",
        )

    except KeyError as exc:
        raise HTTPException(
            status_code=404,
            detail=str(exc),
        ) from exc

    except (
        ValueError,
        PermissionError,
    ) as exc:
        raise HTTPException(
            status_code=409,
            detail=str(exc),
        ) from exc

    return {
        "action": "approved",
        "proposal": proposal,
    }


@router.post(
    "/proposals/{proposal_id}/reject"
)
async def reject_memory_proposal(
    proposal_id: str,
) -> dict:
    orchestrator = _orchestrator()

    try:
        proposal = orchestrator.decide(
            proposal_id,
            approved=False,
            actor_type="user",
            actor_id="local-operator",
        )

    except KeyError as exc:
        raise HTTPException(
            status_code=404,
            detail=str(exc),
        ) from exc

    except (
        ValueError,
        PermissionError,
    ) as exc:
        raise HTTPException(
            status_code=409,
            detail=str(exc),
        ) from exc

    return {
        "action": "rejected",
        "proposal": proposal,
    }

File: frontend/src/api/memory.ts
export type MemoryProposalStatus =
  | "pending"
  | "committed"
  | "rejected";

export interface MemoryProposal {
  proposal_id: string;
  status: MemoryProposalStatus;
  actor_type: string;
  actor_id: string;
  namespace: string;
  scope: string;
  kind: string;
  content: string;
  tags: string[];
  source: string | null;
  policy_decision_id: string | null;
  policy_reason: string | null;
  memory_id: string | null;
  created_at: string;
  updated_at: string;
  decided_by_type: string | null;
  decided_by_id: string | null;
  decided_at: string | null;
}

export interface MemoryProposalList {
  status_filter: string | null;
  count: number;
  proposals: MemoryProposal[];
}

export interface MemoryProposalAuditEvent {
  id: number;
  proposal_id: string;
  action: string;
  actor_type: string;
  actor_id: string;
  detail: Record<string, unknown>;
  timestamp: string;
}

export interface MemoryProposalAudit {
  proposal_id: string;
  count: number;
  events: MemoryProposalAuditEvent[];
}

export async function getMemoryProposals(
  status = "pending",
): Promise<MemoryProposalList> {
  const response = await fetch(
    `/api/memory/proposals?status=${encodeURIComponent(status)}`,
  );

  if (!response.ok) {
    throw new Error(
      `Memory proposals request failed: ${response.status}`,
    );
  }

  return response.json();
}

export async function getMemoryProposalAudit(
  proposalId: string,
): Promise<MemoryProposalAudit> {
  const response = await fetch(
    `/api/memory/proposals/${encodeURIComponent(proposalId)}/audit`,
  );

  if (!response.ok) {
    throw new Error(
      `Memory audit request failed: ${response.status}`,
    );
  }

  return response.json();
}

export async function approveMemoryProposal(
  proposalId: string,
): Promise<void> {
  const response = await fetch(
    `/api/memory/proposals/${encodeURIComponent(proposalId)}/approve`,
    {
      method: "POST",
    },
  );

  if (!response.ok) {
    throw new Error(
      `Memory approval failed: ${response.status}`,
    );
  }
}

export async function rejectMemoryProposal(
  proposalId: string,
): Promise<void> {
  const response = await fetch(
    `/api/memory/proposals/${encodeURIComponent(proposalId)}/reject`,
    {
      method: "POST",
    },
  );

  if (!response.ok) {
    throw new Error(
      `Memory rejection failed: ${response.status}`,
    );
  }
}
